### ISSUE TITLE
file module doesn't create hard link to given target directory

### ISSUE THREAD

**Issue by ericzolf** [ansible/ansible#22738]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
file module

##### ANSIBLE VERSION
```
ansible 2.2.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```

##### CONFIGURATION
none

##### OS / ENVIRONMENT
Fedora 25 (localhost only)

##### SUMMARY
Hard linking with the file module giving a directory as destination doesn't work even though the module recognizes what should be the destination.

##### STEPS TO REPRODUCE

```yaml
- hosts: localhost
  connection: local
  gather_facts: no

  tasks:
    - name: "create subdirectory of /tmp"
      file:
        state: directory
        dest: "/tmp/bla/"
    - name: "create hardlink for file xxx"
      file:
        state: hard
        src: "/tmp/xxx"
        dest: "/tmp/bla/"
```

##### EXPECTED RESULTS

a hard link with the same name as the source file is created in the given directory.

##### ACTUAL RESULTS

```
TASK [create hardlink for file xxx] ********************************************
task path: /tmp/bla.yml:10
Using module file /usr/lib/python2.7/site-packages/ansible/modules/core/files/file.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: myuser
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo ~/.ansible/tmp/ansible-tmp-1489759101.83-10857093995324 `" && echo ansible-tmp-1489759101.83-10857093995324="` echo ~/.ansible/tmp/ansible-tmp-1489759101.83-10857093995324 `" ) && sleep 0'
<127.0.0.1> PUT /tmp/tmpeDY1tP TO /home/myuser/.ansible/tmp/ansible-tmp-1489759101.83-10857093995324/file.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/myuser/.ansible/tmp/ansible-tmp-1489759101.83-10857093995324/ /home/myuser/.ansible/tmp/ansible-tmp-1489759101.83-10857093995324/file.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python2 /home/myuser/.ansible/tmp/ansible-tmp-1489759101.83-10857093995324/file.py; rm -rf "/home/myuser/.ansible/tmp/ansible-tmp-1489759101.83-10857093995324/" > /dev/null 2>&1 && sleep 0'
fatal: [localhost]: FAILED! => {
    "changed": false, 
    "dest": "/tmp/bla/xxx", 
    "failed": true, 
    "gid": 12345, 
    "group": "myuser", 
    "invocation": {
        "module_args": {
            "backup": null, 
            "content": null, 
            "delimiter": null, 
            "dest": "/tmp/bla/", 
            "diff_peek": null, 
            "directory_mode": null, 
            "follow": false, 
            "force": false, 
            "group": null, 
            "mode": null, 
            "original_basename": null, 
            "owner": null, 
            "path": "/tmp/bla/xxx", 
            "recurse": false, 
            "regexp": null, 
            "remote_src": null, 
            "selevel": null, 
            "serole": null, 
            "setype": null, 
            "seuser": null, 
            "src": "/tmp/xxx", 
            "state": "hard", 
            "unsafe_writes": null, 
            "validate": null
        }, 
        "module_name": "file"
    }, 
    "mode": "0664", 
    "msg": "Cannot link, directory exists at destination", 
    "owner": "myuser", 
    "secontext": "unconfined_u:object_r:user_tmp_t:s0", 
    "size": 0, 
    "src": "/tmp/xxx", 
    "state": "hard", 
    "uid": 12345
}
```

The output shows that the file module actually knows where the `dest` should be but seems to use the wrong variable to check for existence.

**Comment by snowsky** [comment]
Interesting case, seems I found something.

**Comment by snowsky** [comment]
Got the same issue, and it seems the variable prev_state is not correctly set.

**Comment by abadger** [comment]
https://github.com/ansible/ansible/pull/22771 has been merged to devel (for the 2.4.0 release). Closing this ticket.
